File: Convert-Sorted-List-to-Binary-Search-Tree/Convert-Sorted-List-to-Binary-Search-Tree.py

‘’’
 Definition for a  binary tree node
 class TreeNode:
     def __init__(self, x):
         self.val = x
         self.left = None
         self.right = None
#
 Definition for singly-linked list.
 class ListNode:
     def __init__(self, x):
         self.val = x
         self.next = None
‘’’
class Solution:
    def buildTree(self,num,l,r):
        if l>r:
            return None
        m=l+(r-l)/2
        root=TreeNode(num[m])
        root.left=self.buildTree(num,l,m-1)
        root.right=self.buildTree(num,m+1,r)
        return root
    # @param head, a list node
    # @return a tree node
    def sortedListToBST(self, head):
        num=[]
        temp=head
        while temp!=None:
            num.append(temp.val)
            temp=temp.next
        return self.buildTree(num,0,len(num)-1)
 

File: Convert-Sorted-List-to-Binary-Search-Tree/Convert-Sorted-List-to-Binary-Search-Tree.cpp

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode *buildTree(ListNode *&head, int l, int r)
    {
        if(l>r)
            return NULL;
        else
        {
            int m=l+(r-l)/2;
            TreeNode *root = new TreeNode(0);
            root->left=buildTree(head,l,m-1);
            root->val = head->val;
            head=head->next;
            root->right=buildTree(head,m+1,r);
            return root;
        }
    }
    TreeNode *sortedListToBST(ListNode *head) {
        ListNode *p=head;
        int n=0;
        while(p)
        {
            n++;
            p=p->next;
        }
        return buildTree(head,0,n-1);
    }
};


 